File: leetcode/2024_7_july/python/04_merge_nodes_in_between_zeros.py
# CODE

from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = head.next
        sum_node = dummy
        while sum_node:
            sum = 0
            while sum_node.val != 0:
                sum += sum_node.val
                sum_node = sum_node.next
            dummy.val = sum
            sum_node = sum_node.next
            dummy.next = sum_node
            dummy = dummy.next
        return head.next

File: leetcode/2024_7_july/c++/04_merge_nodes_in_between_zeros.cpp
// TIME COMPLEXITY
// O(n)

// SPACE COMPLEXITY
// O(1)

// CODE

class Solution {
public:
    ListNode* mergeNodes(ListNode* head) {
        ListNode* dum1= new ListNode(10);
        ListNode* temp=head;
        ListNode* uupp=dum1;
        while(temp->next!=NULL){
            if(temp->val==0){
              temp=temp->next;
              continue;
            } 
            int sum=0;
            while(temp->val!=0){
                sum+=temp->val;
                temp=temp->next;
            }
            uupp->next=new ListNode(sum);
            uupp=uupp->next;
        }
 return dum1->next;
    }
};